commands: add the 'sdk' command to manage SDK.

soburi opened the pull request:
This command can list and install SDK.

Run 'west sdk install' to install the SDK.
Run without any parameter, installing specified by SDK_VERSION in the source tree.
'west sdk' to show installed SDK information.

This command is just a wrapper for SDK's setup command,
but it simplifies the installation process. It will be a good improvement for onboarding first-time users.

--

Since I have no Apple's machine.
Only check on Windows and Linux.

soburi committed:
commands: add the 'sdk' command to manage SDK.

This command can list and install SDK.

Run 'west sdk install' to install the SDK.
Run without any parameter, installing
specified by SDK_VERSION in the source tree.
'west sdk' to show installed SDK information.

This command is just a wrapper for SDK's setup command,
but it simplifies the installation process. It will be a good
improvement for onboarding first-time users.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

pdgendt commented on scripts/west_commands/sdk.py:
```diff
@@ -0,0 +1,368 @@
+                        cmds.extend([optsep + "t", tc])
+                    else:
+                        self.die("toolchain {tc} is not available.\n" +
+                                 "Pelase select list below:\n" +
```
```suggestion
                                 "Please select from the list below:\n" +
```

soburi replied:
Thank you for reviewing.
Fixed it.

pdgendt commented on scripts/requirements-base.txt:
```diff
@@ -18,6 +18,7 @@ canopen
 packaging
 progress
 psutil
+py7zr
```
If this is windows only, maybe?

```suggestion
py7zr; sys_platform == "win32"
```

soburi replied:
I applied your suggestion.

pdgendt commented on scripts/west_commands/sdk.py:
```diff
@@ -0,0 +1,368 @@
+            self.die(f"command \"{' '.join(cmds)}\" failed")
+
+    def list_sdk(self, args, user_args):
+        cmake_script = """
```
Why embed this in a python script, and not make it an actual CMake file?

gmarull replied:
just put the file in a subfolder and open it

soburi replied:
I put the embedded source into scripts/west_commands/sdk/findsdk.cmake.

fabiobaltieri commented on scripts/west_commands/sdk.py:
```diff
@@ -0,0 +1,368 @@
+            ),
+        )
+
+        parser.add_argument(
+            "--install",
+            default=None,
+            dest="sdk_version",
+            metavar="SDK_VER",
+            help="Install the selected version of Zephyr SDK. "
+            'If you set this option to "latest", select the latest version.',
```
there's a `SDK_VERSION` file at the top of the main repository, that is the one used by CI so it's in a way the "current" one for the current checkout, how about a way of picking that version? once it's deployed it simplify https://github.com/zephyrproject-rtos/action-zephyr-setup/blob/main/action.yml a lot

soburi replied:
I changed it to refer to the SDK_VERSION by default.
Also, Add `--install-dir` option for considering the CI use case.
This option allows the modification of the SDK install destination.

gmarull commented:
would it be worth updating GSG to use this tool?

soburi commented:
> would it be worth updating GSG to use this tool?

That's the purpose of this PR.
If you all like it, we'd love to see it reflected.

pdgendt commented on scripts/west_commands/sdk.py:
```diff
@@ -0,0 +1,518 @@
+
+    def fetch_sdk_info(self):
+        sorted_sdks = []
+        cmds = [
+            "cmake",
+            "-P",
+            os.path.join(os.path.dirname(__file__), "sdk", "listsdk.cmake"),
+        ]
```
You could make use of `zcmake.py`.

soburi replied:
I fixed it. Thanks.

pdgendt commented on scripts/west_commands/sdk.py:
```diff
@@ -0,0 +1,518 @@
+        else:
+            self.die(f"Unsupported system: {system}")
+
+        if machine == "aarch64":
```
Works on macOS if I add `arm64` (Apple silicon)

```suggestion
        if machine in ["aarch64", "arm64"]:
```

soburi replied:
Thank you for verifying, and I am sorry for taking your time.
Fixed it.

fabiobaltieri commented on scripts/west_commands/sdk/listsdk.cmake:
+cc @tejlmand for the cmake file

tejlmand commented on scripts/west_commands/sdk.py:
```diff
@@ -0,0 +1,525 @@
+            metavar="TC",
+            nargs="*",
+            help="specifies the installation of toolchains. "
+            "If neither is specified, all toolchains will be installed.",
```
I assume this refers to cross toolchains for specific architectures, such as xtensa, arm, espressif, and so on.

However, I don't find that clear from this text.

Also the user of `neither` here is not clear if reading the help text when it's printed:
```
positional arguments:
  {list,install}        select a subcommand. 'list' is the default.

options:
  -h, --help            show this help message and exit
  -v [SDK_VER], --version [SDK_VER]
                        specifies the install version of Zephyr SDK.If not specified, the install version detects from ${ZEPHYR_BASE}/SDK_VERSION file.
  -d DIR, --install-dir DIR
                        specifies the install destination directory.
  -i, --interactive     launches installer in interactive mode. --toolchains, --no-toolchains and --no-hosttools will be ignored if this option enabled.
  -t [TC ...], --toolchains [TC ...]
                        specifies the installation of toolchains. If neither is specified, all toolchains will be installed.
  -T, --no-toolchains   do not install toolchains. --toolchains will be ignored if this option is enabled.
  -H, --no-hosttools    do not install host-tools.
  --personal-access-token PERSONAL_ACCESS_TOKEN
                        specify GitHub personal access token.
  --api-url API_URL     specifies the GitHub releases API endpoint for finding SDKs.

```

please be clearer on what exactly  neither refers to.

soburi replied:
I changed the `--toolchains` to a constraint that takes one or more arguments.
This change eliminates the zero-length arguments case and makes it clear that all toolchains are installed when the toolchains option is not specified, so I state so.

soburi replied:
> So please split list and install into two distinct sub-commands.

Fixed it.

tejlmand commented on scripts/west_commands/sdk/listsdk.cmake:
I would like to avoid this file if possible.

I believe we can rework https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/modules/FindZephyr-sdk.cmake a bit to support this use-case as well.

soburi replied:
Because `FindZephyr-sdk.cmake` is executed in a special context of `find_package`, it is better not to give it too many meanings, so I think it is better to have this file.
However, to avoid duplicate implementation, I separated the part extracted from `FindZephyr-sdk.cmake` into `zephyr_sdk.cmake` and made it shared.

tejlmand replied:
with #77162 this code can be changed into:
```
cmake_minimum_required(VERSION 3.20)                                                                                                                                                                                                                           
                                                                                                                                                                                                                                                               
set(CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/modules)                                                                                                                                                                                                       
                                                                                                                                                                                                                                                               
find_package(Zephyr-sdk COMPONENTS LIST)
```

Which will print:
```
-- Zephyr-sdk, version=0.16.8, dir=/opt/zephyr-sdk-0.16.8                                                                                                                                                                                                      
-- Zephyr-sdk, version=0.15.0, dir=/home/tora/zephyr-sdk-0.15.0 
```

tejlmand replied:
Also move this script into `<zephyr-base>/cmake` because that's where we generally place CMake scripts.

tejlmand reviewed (changes requested):
I like the feature proposal, but I think the two sub-commands, `list` and `install` should be made more independent with their own arguments.

Most arguments (if not all arguments) only make sense for `install` but can still be given for `list`.

For example what does `--install-dir` mean when specified together with `list` ?
Technically it could mean `Look only for installed toolchains in this location`.

So please split `list` and `install` into two distinct sub-commands.

fabiobaltieri commented:
@tejlmand can you take another look?

gmarull reviewed (dismissed):
update GSG once merged

kartben commented on scripts/west_commands/sdk.py:
```diff
@@ -0,0 +1,553 @@
+                    if sha256 != digest:
+                        raise Exception(f"sha256 mismatched: {sha256}:{digest}")
+
+                # extract archive file
+                self.inf(f"Extract: {file.name}")
+                if file.name.endswith(".tar.xz"):
+                    with tarfile.open(file.name, mode="r:xz") as archive:
+                        archive.extractall(path=tempdir)
+                elif file.name.endswith(".tar.gz"):
+                    with tarfile.open(file.name, mode="r:gz") as archive:
+                        archive.extractall(path=tempdir)
+                elif file.name.endswith(".7z"):
+                    with SevenZipFile(file.name, mode="r") as archive:
+                        archive.extractall(path=tempdir)
+                elif file.name.endswith(".zip"):
+                    with ZipFile(file.name, mode="r") as archive:
+                        archive.extractall(path=tempdir)
+                else:
+                    raise Exception("Unknown file extension.")
```
can't you use patool? it feels weird to not rely on an existing well established module to deal with this kind of boilerplate

kartben commented on cmake/modules/zephyr_sdk.cmake:
```diff
@@ -0,0 +1,64 @@
+  list(REMOVE_DUPLICATES Zephyr-sdk_CONSIDERED_VERSIONS)
+  list(SORT Zephyr-sdk_CONSIDERED_VERSIONS COMPARE NATURAL ORDER DESCENDING)
+
+  # Loop over each found Zepher SDK version until one is found that is compatible.
```
```suggestion
  # Loop over each found Zephyr SDK version until one is found that is compatible.
```

kartben commented on scripts/west_commands/sdk.py:
```diff
@@ -0,0 +1,553 @@
+                zephyr_base = os.path.join(os.path.dirname(__file__), "..", "..")
+
+            if not os.path.exists(os.path.join(zephyr_base, "SDK_VERSION")):
+                self.die(f'{os.path.join(zephyr_base, "SDK_VERSION")} not exists.')
```
```suggestion
                self.die(f'{os.path.join(zephyr_base, "SDK_VERSION")} does not exist.')
```

kartben commented on scripts/west_commands/sdk.py:
```diff
@@ -0,0 +1,553 @@
+        install_args_parser.add_argument(
+            "-t",
+            "--toolchains",
+            metavar="TC",
```
```suggestion
            metavar="toolchain_name",
```

kartben commented on scripts/west_commands/sdk.py:
```diff
@@ -0,0 +1,553 @@
+            "--toolchains",
+            metavar="TC",
+            nargs="+",
+            help="specifies the installation of toolchains. "
```
```suggestion
            help="toolchain(s) to install (e.g. 'arm-zephyr-eabi'). "
```

kartben commented on scripts/west_commands/sdk.py:
```diff
@@ -0,0 +1,553 @@
+            help="do not install host-tools.",
+        )
+        install_args_parser.add_argument(
+            "--personal-access-token", help="specify GitHub personal access token."
```
```suggestion
            "--personal-access-token", help="GitHub personal access token."
```

kartben commented on scripts/west_commands/sdk.py:
```diff
@@ -0,0 +1,553 @@
+        install_args_parser.add_argument(
+            "--api-url",
+            default="https://api.github.com/repos/zephyrproject-rtos/sdk-ng/releases",
+            help="specifies the GitHub releases API endpoint for finding SDKs.",
```
```suggestion
            help="GitHub releases API endpoint used to look for Zephyr SDKs.",
```

kartben commented on scripts/west_commands/sdk.py:
```diff
@@ -0,0 +1,553 @@
+            "--install-dir",
+            default=None,
+            metavar="DIR",
+            help="specifies the install destination directory.",
```
```suggestion
            help="SDK install destination directory.",
```

kartben commented on scripts/west_commands/sdk.py:
```diff
@@ -0,0 +1,553 @@
+            default=None,
+            nargs="?",
+            metavar="SDK_VER",
+            help="specifies the install version of Zephyr SDK. "
```
```suggestion
            help="version of the Zephyr SDK to install. "
```

kartben commented on scripts/west_commands/sdk.py:
```diff
@@ -0,0 +1,553 @@
+
+                Run 'west sdk install' to install Zephyr SDK.
+
+                Set --version option to install the specified version of the SDK.
```
```suggestion
                Set --version option to install a specific version of the SDK.
```

kartben commented on scripts/west_commands/sdk.py:
```diff
@@ -0,0 +1,553 @@
+                the already installed SDK will be used regardless of the settings of
+                --install-dir.
+
+                Typically, Zephyr SDK archives contain only a directory named zephyr-sdk-<version>
```
```suggestion
                Typically, Zephyr SDK archives contain only one directory named zephyr-sdk-<version>
```

kartben commented on scripts/west_commands/sdk.py:
```diff
@@ -0,0 +1,553 @@
+            else:
+                zephyr_base = os.path.join(os.path.dirname(__file__), "..", "..")
+
+            if not os.path.exists(os.path.join(zephyr_base, "SDK_VERSION")):
```
I would recommend switching to pathlib across the entire file

```suggestion
            if not (Path(zephyr_base) / "SDK_VERSION").exists():
```

soburi replied:
Fixed it. 
I revisited this point overall.

kartben commented on scripts/west_commands/sdk.py:
```diff
@@ -0,0 +1,553 @@
+            else:
+                ext = ".tar.xz"
+
+        return "zephyr-sdk-" + version + "_" + osname + "-" + arch + "_minimal" + ext
```
nit: use `f"..."`

soburi replied:
Fixed it.

kartben commented on scripts/west_commands/sdk.py:
```diff
@@ -0,0 +1,553 @@
+            self.die("No Zephyr SDK installed.")
+
+        for k, v in sdk_info.items():
+            self.inf(k + ":")
+            self.inf("  path: " + v["path"])
```
you may want to be more consistent in using formatted strings instead of concatenation.

soburi replied:
Fixed it.

kartben reviewed (changes requested):
Can you please also update the files in `scripts/west_commands/completion`. I would be ok with it being in a different PR though, just afraid that this will be forgotten if we don't do it right away.

Thanks a lot for this work, this is super cool ❤️

soburi commented on scripts/west_commands/sdk.py:
```diff
@@ -0,0 +1,553 @@
+                Run 'west sdk install' to install Zephyr SDK.
+
+                Set --version option to install the specified version of the SDK.
+                If not specified, the install version detects from "${ZEPHYR_BASE}/SDK_VERSION file.
```
I corrected all points regarding wording. Thank you.

kartben commented:
> Can you please also update the files in `scripts/west_commands/completion`. I would be ok with it being in a different PR though, just afraid that this will be forgotten if we don't do it right away.
> 
> Thanks a lot for this work, this is super cool ❤️ 


@soburi you may not have seen this in the flow of comments. Thanks!

soburi committed:
scripts: west_commands: completion: bash: Add support for `west sdk`

Add a feature for completion of the `west sdk` command.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi commented:
> > Can you please also update the files in `scripts/west_commands/completion`. I would be ok with it being in a different PR though, just afraid that this will be forgotten if we don't do it right away.
> > Thanks a lot for this work, this is super cool ❤️
> 
> @soburi you may not have seen this in the flow of comments. Thanks!

I'm just starting to think about it, so I've just made a draft for work management purposes.

https://github.com/zephyrproject-rtos/zephyr/pull/76906

kartben commented:
Oh, right. I guess I should have clarified that I think initially it would be fine if you just propose completion for the initial `sdk` token, like it's done for all the other commands. Essentially just making sure that people typing `west <TAB>` know about the existence of the `sdk` command. More complete suggestions is optional/second step IMO. Thanks again!

soburi committed:
scripts: west_commands: completion: zsh: Add minimum support for `sdk`

Add minimum support for the `west sdk` command.
This only provides completion for the `sdk` command itself.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
scripts: west_commands: completion: fash: Add support for `west sdk`

Add a feature for completion of the `west sdk` command.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi commented:
@kartben 
> Oh, right. I guess I should have clarified that I think initially it would be fine if you just propose completion for the initial `sdk` token, like it's done for all the other commands. Essentially just making sure that people typing `west <TAB>` know about the existence of the `sdk` command. More complete suggestions is optional/second step IMO. Thanks again!

I added a minimal implementation of completion.

tejlmand reviewed (changes requested):
Providing an install dir which is almost at top, like `/opt` is not working.

Seems a path separator is missing after `/tmp/` cause it tries to create a new folder directly at `/`.

Example:
```
$ west sdk install -v 0.15.1 --install-dir /opt/
Fetching Zephyr SDK list...
Fetching sha256...
Downloading https://github.com/zephyrproject-rtos/sdk-ng/releases/download/v0.15.1/zephyr-sdk-0.15.1_linux-x86_64_minimal.tar.gz...
FATAL ERROR: [Errno 13] Permission denied: '/tmpb1glrunv'
```

tejlmand commented on scripts/west_commands/sdk.py:
```diff
@@ -0,0 +1,528 @@
+        )
+
+        install_args_parser.add_argument(
+            "-v",
```
please remove the short-arg `-v` because `west` itself has already established the meaning of `-v` as:
```
$ west -h
usage: west [-h] [-z ZEPHYR_BASE] [-v] [-V] <command> ...

The Zephyr RTOS meta-tool.

optional arguments:
...
  -v, --verbose         Display verbose output. May be given multiple times
                        to increase verbosity.
```
so although the west `-v` and the `sdk` command argument `-v` must be placed at different locations, then it's still slightly confusing that the `-v` can mean different things depending on placement, for example:
```
$ west -v sdk install -v 0.16.0
        ^              ^-- means version
        |----------------- means verbose
```
and yes, the above is indeed a valid command to run :slightly_smiling_face:

soburi replied:
I don't think this is a bad specification, but since specifying the version is a case where there is a clear intention, I think it would be better to explicitly specify it with the --version. I have removed the -v option.

tejlmand commented on cmake/modules/FindZephyr-sdk.cmake:
```diff
@@ -61,53 +63,7 @@ if(("zephyr" STREQUAL ${ZEPHYR_TOOLCHAIN_VARIANT}) OR
-      # back to the minimum version so that it is displayed in the error text.
-      find_package(Zephyr-sdk ${Zephyr-sdk_FIND_VERSION} REQUIRED CONFIG PATHS ${zephyr_sdk_search_paths})
-    endif()
+    find_zephyr_sdk(0)
```
@soburi  Created a cleanup and enhancement PR for FindZephyr-sdk module here: https://github.com/zephyrproject-rtos/zephyr/pull/77162

See also this comment: https://github.com/zephyrproject-rtos/zephyr/pull/76283#discussion_r1719709221

soburi replied:
Thank you for updating `FindZephyr-sdk.cmake`.
I incorporated it and succeeded in working fine.
Since this PR needs to merge after #77162 is merged. 
I put DNM mark on this PR.

soburi commented on scripts/west_commands/sdk.py:
```diff
@@ -0,0 +1,528 @@
+            "--toolchains",
+            metavar="toolchain_name",
+            nargs="+",
+            help="toolchain(s) to install (e.g. 'arm-zephyr-eabi'). "
+            "If this option is not given, all toolchains will be installed.",
```
Thank you for your suggestion.
I applied it.

tejlmand replied:
I don't see an update.

Did you forget to push ?

soburi replied:
I accidentally did a rebase and couldn't push it. Sorry.
Fixed it.

soburi commented on scripts/west_commands/sdk.py:
```diff
@@ -0,0 +1,528 @@
+            "--toolchains",
+            metavar="toolchain_name",
+            nargs="+",
+            help="toolchain(s) to install (e.g. 'arm-zephyr-eabi'). "
+            "If this option is not given, all toolchains will be installed.",
```
SDK 0.15.2 and later versions have the sdk_toolchains file, so if this file is present, I made the output of `sdk list` display toolchains that are not installed.
I have decided that displaying help is unpreferable because it requires retrieving information via the network.

soburi commented:
@tejlmand 

> Providing an install dir which is almost at top, like `/opt` is not working.
> 
> Seems a path separator is missing after `/tmp/` cause it tries to create a new folder directly at `/`.
> 
> Example:
> 
> ```
> $ west sdk install -v 0.15.1 --install-dir /opt/
> Fetching Zephyr SDK list...
> Fetching sha256...
> Downloading https://github.com/zephyrproject-rtos/sdk-ng/releases/download/v0.15.1/zephyr-sdk-0.15.1_linux-x86_64_minimal.tar.gz...
> FATAL ERROR: [Errno 13] Permission denied: '/tmpb1glrunv'
> ```

In the current specification, --install-dir specifies the complete path to where the SDK will be installed. In other words, the zephyr-sdk-x.y.z part in the archive is also renamed and placed in the specified path.
This was done to avoid creating multiple directory specification options and to accommodate use cases expected in CI.
However, since the use case you pointed out is also anticipated, we have added the --install-base option to specify the extraction destination.

tejlmand commented:
> In the current specification, --install-dir specifies the complete path to where the SDK will be installed. In other words, the zephyr-sdk-x.y.z part in the archive is also renamed and placed in the specified path.

In that case I would say this help text is wrong (or as minimum ambiguous and needs to be improved):
https://github.com/zephyrproject-rtos/zephyr/blob/f04e6db67c2757f5bc3bc9ef061af9f97612ed69/scripts/west_commands/sdk.py#L128

because when I examine the tar-ball, then the top-level directory in there is `zephyr-sdk-x.y.z` like this:
```
$ tar --exclude="*/*" -tf ~/Downloads/zephyr-sdk-0.16.8_linux-x86_64.tar.xz 
zephyr-sdk-0.16.8/
```

so based on the linked help text, I would assume that the top-level folder in that tar-ball, i.e. `zephyr-sdk-0.16.8` is placed under `/opt` when given `--install-dir /opt`.

tejlmand commented on scripts/west_commands/completion/west-completion.bash:
```diff
@@ -1142,6 +1142,52 @@ __comp_west_twister()
+	"
+
+	local other_opts="
+		--version -v
```
As short form has been removed, then this should be updated.
```suggestion
		--version
```

Same probably goes for other completion handlers.

Btw, thank for providing completion handlers from the start (y)

soburi replied:
> As short form has been removed, then this should be updated.

I forgot about this one too. Thank you for pointing out.

> Btw, thank for providing completion handlers from the start 👍
I have finished bash and fish, but zsh requires some consideration, so I only included the bare minimum.
I'll update this soon.

soburi commented:
@tejlmand 
https://github.com/zephyrproject-rtos/zephyr/pull/76283#issuecomment-2296469140

> In that case I would say this help text is wrong (or as minimum ambiguous and needs to be improved):

Thank you for checking.
It's a confusing situation, so I've tried to explain it with a concrete example.
I hope this makes it clearer.

@kartben 
It would be helpful if you could look at it from the standpoint of readability.

tejlmand commented:
> > In that case I would say this help text is wrong (or as minimum ambiguous and needs to be improved):
>
> Thank you for checking.
> It's a confusing situation, so I've tried to explain it with a concrete example.
> I hope this makes it clearer.

thanks, and yes, the new help text makes it clearer.
It was actually because of the old help text that I provided `--install-dir /opt`, so good we got that sentence made clearer.

I'll leave it for @kartben to review the help text in depth. To me it looks clear enough.
